ui/app/services/token.js

```diff
@@ -115,7 +120,67 @@ export default class TokenService extends Service {
     this.fetchSelfToken.cancelAll({ resetState: true });
     this.fetchSelfTokenPolicies.cancelAll({ resetState: true });
     this.fetchSelfTokenAndPolicies.cancelAll({ resetState: true });
+    this.monitorTokenTime.cancelAll({ resetState: true });
   }
+
+  kickoffTokenTTLMonitoring() {
+    this.monitorTokenTime.perform();
+  }
+
+  @task(function* () {
+    while (this.selfToken) {
+      const diff = new Date(this.selfToken.expirationTime) - new Date();
+      // Let the user know at the 10 minute mark,
+      // or any time they refresh with under 10 minutes left
+      if (diff < 1000 * 60 * MINUTES_LEFT_AT_WARNING) {
+        const existingNotification = this.flashMessages.queue?.find(
+          (m) => m.title === EXPIRY_NOTIFICATION_TITLE
+        );
+        // For the sake of updating the "time left" message, we keep running the task down to the moment of expiration
+        if (diff > 0) {
+          if (existingNotification) {
+            existingNotification.set(
+              'message',
+              `Your token access expires ${moment(
+                this.selfToken.expirationTime
+              ).fromNow()}`
+            );
+          } else {
+            if (!this.expirationNotificationDismissed) {
+              this.flashMessages.add({
+                title: EXPIRY_NOTIFICATION_TITLE,
+                message: `Your token access expires ${moment(
+                  this.selfToken.expirationTime
+                ).fromNow()}`,
+                type: 'error',
+                destroyOnClick: false,
+                sticky: true,
+                customCloseAction: () => {
+                  this.set('expirationNotificationDismissed', true);
+                },
+                customAction: {
+                  label: 'Re-authenticate',
+                  action: () => {
+                    this.router.transitionTo('settings.tokens');
+                  },
+                },
+              });
+            }
+          }
+        } else {
+          if (existingNotification) {
+            existingNotification.setProperties({
+              title: 'Your access has expired',
+              message: `Your token will need to be re-authenticated`,
+            });
+          }
+          this.monitorTokenTime.cancelAll(); // Stop updating time left after expiration
+        }
+      }
+      yield timeout(1000);
+    }
+  })
+  monitorTokenTime;
```

Comment:
This is using a syntax common to Nomad's front-end, inherited from [Ember Concurrency](http://ember-concurrency.com/docs/introduction/). There were a few attempts to handle this using native JS async/await but, despite the large block size, this ended up feeling more streamlined.